tests/study_group/test_study_group.py

```diff
@@ -56,3 +56,8 @@ def test_field_over_char_limit(self, persist_group, over_char_limit):
         persist_group.field = over_char_limit
         with pytest.raises(ValidationError):
             persist_group.save()
+
+    def test_join_get_members_count(self, persist_group):
+        for i in range(1, persist_group.capacity):
```

Comment:
Here the static `1` in `range` is only technical-
It is used so the current `i` value will match the expected amount of members in the group in each iteration.
The behavior of attempting to join a group of full capacity was already tested in different test

Reply:
Do we really need to test it for every iteration? I tend to agree with Omer here that 0, 1, `max_capacity`, and `max_capacity + 1` might be enough

Reply:
I guess that parametrizing the test does make more sense, i will change it.
Why does `max_capacity + 1` case needs to be checked in this test?
There is already a test that checks the behavior of joining a group at max capacity

Reply:
> I guess that parametrizing the test does make more sense, i will change it. Why does `max_capacity + 1` case needs to be checked in this test? There is already a test that checks the behavior of joining a group at max capacity

Correct, no need for a duplicate test

> Now that I think about it, one of the reasons for checking every iteration was that i didnt want the test to be dependent on a hardcoded `max_capacity` value inside the parametrize decorator. I can't access the capacity value of the group fixture there as far as i know.

You can always pass it as another parameter, something like `@pytest.mark.parametrize("capacity, max_capacity", [(1, 5)])`

Reply:
> > I guess that parametrizing the test does make more sense, i will change it. Why does `max_capacity + 1` case needs to be checked in this test? There is already a test that checks the behavior of joining a group at max capacity
> 
> Correct, no need for a duplicate test
> 
> > Now that I think about it, one of the reasons for checking every iteration was that i didnt want the test to be dependent on a hardcoded `max_capacity` value inside the parametrize decorator. I can't access the capacity value of the group fixture there as far as i know.
> 
> You can always pass it as another parameter, something like `@pytest.mark.parametrize("capacity, max_capacity", [(1, 5)])`

What I meant is, by doing that, If someone changes this fixture's capacity field, 5 would no longer be the max capacity.
If for example, someone lowers the fixture's capacity field, the test will fail because more users are attempted to be added than the capacity.

I guess its not likely that a capacity of a fixture would be modified, so in a way it does make more sense to parametrize  it. Will change it.

Reply:
I'd suggest an idea, and a solution:
1. The fixture should ease your tests, not prevent you from writing good tests. if it does, it mean something is wrong, either the test, the fixture, or the need for the fixture.
2. checkout `fixture_factory`, with that you can use a fixture that gets arguments, therefor modifying it. for instance, you can have a fixture that create a user where 95% of the fields have static data, but it requires the `name` and `max_capacity` arguments (with defaults in case it doesn't get them), then you can create copies of this user, with different details, using the fixture, on each iteration of the test.

Reply:
Done

Reply:
Great!